File: utttpy/game/ultimate_tic_tac_toe.py
from __future__ import annotations

from typing import List, Optional

from utttpy.game.action import Action
from utttpy.game.constants import (
    STATE_SIZE,
    NEXT_SYMBOL_STATE_INDEX,
    CONSTRAINT_STATE_INDEX,
    UTTT_RESULT_STATE_INDEX,
    X_STATE_VALUE,
    O_STATE_VALUE,
    DRAW_STATE_VALUE,
    UNCONSTRAINED_STATE_VALUE,
)


class UltimateTicTacToe:

    def __init__(self, state: Optional[bytearray] = None):
        if state:
            self.state = state
        else:
            self.state = bytearray(STATE_SIZE)
            self.state[NEXT_SYMBOL_STATE_INDEX] = X_STATE_VALUE
            self.state[CONSTRAINT_STATE_INDEX] = UNCONSTRAINED_STATE_VALUE

    def clone(self) -> UltimateTicTacToe:
        return UltimateTicTacToe(state=self.state.copy())

    def is_equal_to(self, uttt: UltimateTicTacToe) -> bool:
        return self.state == uttt.state

    def execute(self, action: Action, verify: bool = True) -> None:
        if verify:
            if self.is_terminated():
                raise UltimateTicTacToeError("supergame is terminated")
            self._verify_state()
            self._verify_action(action=action)
        self.state[action.index] = action.symbol
        self._update_supergame_result(symbol=action.symbol, index=action.index)
        self._toggle_next_symbol()
        self._set_next_constraint(index=action.index)
        if verify:
            self._verify_state()

    def get_legal_actions(self) -> List[Action]:
        return [
            Action(symbol=self.next_symbol, index=legal_index)
            for legal_index in self.get_legal_indexes()
        ]

    def get_legal_indexes(self) -> List[int]:
        if self.is_terminated():
            return []
        if self.is_unconstrained():
            indexes = []
            for i, s in enumerate(self.state[81:90]):
                if not s:
                    indexes.extend(self._get_empty_indexes(subgame=i))
        else:
            indexes = self._get_empty_indexes(subgame=self.constraint)
        return indexes

    @property
    def next_symbol(self) -> int:
        return self.state[NEXT_SYMBOL_STATE_INDEX]

    @property
    def constraint(self) -> int:
        return self.state[CONSTRAINT_STATE_INDEX]

    @property
    def result(self) -> int:
        return self.state[UTTT_RESULT_STATE_INDEX]

    def is_next_symbol_X(self) -> bool:
        return self.next_symbol == X_STATE_VALUE

    def is_next_symbol_O(self) -> bool:
        return self.next_symbol == O_STATE_VALUE

    def is_constrained(self) -> bool:
        return 0 <= self.constraint < 9

    def is_unconstrained(self) -> bool:
        return self.constraint == UNCONSTRAINED_STATE_VALUE

    def is_terminated(self) -> bool:
        return bool(self.result)

    def is_result_X(self) -> bool:
        return self.result == X_STATE_VALUE

    def is_result_O(self) -> bool:
        return self.result == O_STATE_VALUE

    def is_result_draw(self) -> bool:
        return self.result == DRAW_STATE_VALUE

    def _get_empty_indexes(self, subgame: int) -> List[int]:
        offset = subgame * 9
        return [
            i + offset for i, s in enumerate(self.state[offset : offset + 9]) if not s
        ]

    def _is_winning_position(self, symbol: int, subgame: int) -> bool:
        state = self.state
        offset = subgame * 9
        return (
            (
                symbol == state[offset + 4] and
                (
                    symbol == state[offset + 0] == state[offset + 8] or
                    symbol == state[offset + 2] == state[offset + 6] or
                    symbol == state[offset + 1] == state[offset + 7] or
                    symbol == state[offset + 3] == state[offset + 5]
                )
            )
            or
            (
                symbol == state[offset + 0] and
                (
                    symbol == state[offset + 1] == state[offset + 2] or
                    symbol == state[offset + 3] == state[offset + 6]
                )
            )
            or
            (
                symbol == state[offset + 8] and
                (
                    symbol == state[offset + 2] == state[offset + 5] or
                    symbol == state[offset + 6] == state[offset + 7]
                )
            )
        )

    def _is_full(self, subgame: int) -> bool:
        offset = subgame * 9
        return all(self.state[offset : offset + 9])

    def _update_supergame_result(self, symbol: int, index: int) -> None:
        supergame_updated = False
        subgame = index // 9
        if self._is_winning_position(symbol=symbol, subgame=subgame):
            self.state[81 + subgame] = symbol
            supergame_updated = True
        elif self._is_full(subgame=subgame):
            self.state[81 + subgame] = DRAW_STATE_VALUE
            supergame_updated = True
        if supergame_updated:
            if self._is_winning_position(symbol=symbol, subgame=9):
                self.state[UTTT_RESULT_STATE_INDEX] = symbol
            elif self._is_full(subgame=9):
                self.state[UTTT_RESULT_STATE_INDEX] = DRAW_STATE_VALUE

    def _toggle_next_symbol(self) -> None:
        if self.is_next_symbol_X():
            self.state[NEXT_SYMBOL_STATE_INDEX] = O_STATE_VALUE
        elif self.is_next_symbol_O():
            self.state[NEXT_SYMBOL_STATE_INDEX] = X_STATE_VALUE

    def _set_next_constraint(self, index: int) -> None:
        next_subgame = index % 9
        if self.state[81 + next_subgame]:
            self.state[CONSTRAINT_STATE_INDEX] = UNCONSTRAINED_STATE_VALUE
        else:
            self.state[CONSTRAINT_STATE_INDEX] = next_subgame

    def _verify_state(self) -> None:
        self._verify_supergame()
        self._verify_subgames()
        self._verify_constraint()

    def _verify_supergame(self) -> None:
        x_w = self._is_winning_position(symbol=X_STATE_VALUE, subgame=9)
        o_w = self._is_winning_position(symbol=O_STATE_VALUE, subgame=9)
        full = self._is_full(subgame=9)
        if x_w and o_w:
            raise UltimateTicTacToeError("X and O have winning positions on supergame")
        if x_w and not self.is_result_X():
            raise UltimateTicTacToeError("X won supergame, but result is not updated")
        if o_w and not self.is_result_O():
            raise UltimateTicTacToeError("O won supergame, but result is not updated")
        if full and not self.is_result_draw() and not (x_w or o_w):
            raise UltimateTicTacToeError("DRAW on supergame, but result is not updated")

    def _verify_subgames(self) -> None:
        for subgame in range(0, 9):
            x_w = self._is_winning_position(symbol=X_STATE_VALUE, subgame=subgame)
            o_w = self._is_winning_position(symbol=O_STATE_VALUE, subgame=subgame)
            full = self._is_full(subgame=subgame)
            if x_w and o_w:
                raise UltimateTicTacToeError(f"X and O have winning positions on subgame={subgame}")
            if x_w and self.state[81 + subgame] != X_STATE_VALUE:
                raise UltimateTicTacToeError(f"X won subgame={subgame}, but supergame is not updated")
            if o_w and self.state[81 + subgame] != O_STATE_VALUE:
                raise UltimateTicTacToeError(f"O won subgame={subgame}, but supergame is not updated")
            if full and self.state[81 + subgame] != DRAW_STATE_VALUE and not (x_w or o_w):
                raise UltimateTicTacToeError(f"DRAW on subgame={subgame}, but supergame is not updated")

    def _verify_constraint(self) -> None:
        if not (self.is_constrained() or self.is_unconstrained()):
            raise UltimateTicTacToeError(f"invalid constraint={self.constraint}")
        if self.is_constrained() and self.state[81 + self.constraint]:
            raise UltimateTicTacToeError(f"constraint={self.constraint} points to terminated subgame")

    def _verify_action(self, action: Action) -> None:
        illegal_action = f"Illegal {action} - "
        if self.is_next_symbol_X() and not action.is_symbol_X():
            raise UltimateTicTacToeError(illegal_action + "next move belongs to X")
        if self.is_next_symbol_O() and not action.is_symbol_O():
            raise UltimateTicTacToeError(illegal_action + "next move belongs to O")
        if not (0 <= action.index < 81):
            raise UltimateTicTacToeError(illegal_action + "index outside the valid range")
        if self.is_constrained() and self.constraint != action.index // 9:
            raise UltimateTicTacToeError(illegal_action + f"violated constraint={self.constraint}")
        if self.state[81 + action.index // 9]:
            raise UltimateTicTacToeError(illegal_action + "index from terminated subgame")
        if self.state[action.index]:
            raise UltimateTicTacToeError(illegal_action + "index is already taken")

    def __str__(self):
        state_values_map = {
            X_STATE_VALUE: 'X',
            O_STATE_VALUE: 'O',
            DRAW_STATE_VALUE: '=',
            0: '-',
        }
        subgames = [state_values_map[s] for s in self.state[0:81]]
        supergame = [state_values_map[s] for s in self.state[81:90]]
        if not self.is_terminated():
            for legal_index in self.get_legal_indexes():
                subgames[legal_index] = '•'
            if self.is_constrained():
                supergame[self.constraint] = '•'
            elif self.is_unconstrained():
                supergame = ['•' if s == '-' else s for s in supergame]
        sb = lambda l, r: ' '.join(subgames[l : r + 1])
        sp = lambda l, r: ' '.join(supergame[l : r + 1])
        subgames = [
            '    0 1 2   3 4 5   6 7 8',
            '  0 ' + sb(0, 2) + ' │ ' + sb(9, 11) + ' │ ' + sb(18, 20),
            '  1 ' + sb(3, 5) + ' │ ' + sb(12, 14) + ' │ ' + sb(21, 23),
            '  2 ' + sb(6, 8) + ' │ ' + sb(15, 17) + ' │ ' + sb(24, 26),
            '    ' + '—' * 21,
            '  3 ' + sb(27, 29) + ' │ ' + sb(36, 38) + ' │ ' + sb(45, 47),
            '  4 ' + sb(30, 32) + ' │ ' + sb(39, 41) + ' │ ' + sb(48, 50),
            '  5 ' + sb(33, 35) + ' │ ' + sb(42, 44) + ' │ ' + sb(51, 53),
            '    ' + '—' * 21,
            '  6 ' + sb(54, 56) + ' │ ' + sb(63, 65) + ' │ ' + sb(72, 74),
            '  7 ' + sb(57, 59) + ' │ ' + sb(66, 68) + ' │ ' + sb(75, 77),
            '  8 ' + sb(60, 62) + ' │ ' + sb(69, 71) + ' │ ' + sb(78, 80),
        ]
        supergame = [
            '  ' + sp(0, 2),
            '  ' + sp(3, 5),
            '  ' + sp(6, 8),
        ]
        subgames = '\n'.join(subgames)
        supergame = '\n'.join(supergame)
        next_symbol = state_values_map[self.next_symbol]
        constraint = 'None' if self.is_unconstrained() else str(self.constraint)
        result = 'None'
        if self.is_result_X():
            result = 'X_WON'
        elif self.is_result_O():
            result = 'O_WON'
        elif self.is_result_draw():
            result = 'DRAW'
        output = '{cls}(\n'
        output += '  subgames:\n{subgames}\n'
        if not self.is_terminated():
            output += '  next_symbol: {next_symbol}\n'
            output += '  constraint: {constraint}\n'
        output += '  supergame:\n{supergame}\n'
        output += '  result: {result}\n)'
        output = output.format(
            cls=self.__class__.__name__,
            subgames=subgames,
            supergame=supergame,
            next_symbol=next_symbol,
            constraint=constraint,
            result=result,
        )
        return output


class UltimateTicTacToeError(Exception):
    pass

File: website/src/game/UltimateTicTacToe.js
import {
  STATE_SIZE,
  NEXT_SYMBOL_STATE_INDEX,
  CONSTRAINT_STATE_INDEX,
  UTTT_RESULT_STATE_INDEX,
  X_STATE_VALUE,
  O_STATE_VALUE,
  DRAW_STATE_VALUE,
  UNCONSTRAINED_STATE_VALUE,
} from './constants.js';
import Action from './Action.js';
import UltimateTicTacToeError from './UltimateTicTacToeError.js';

export default class UltimateTicTacToe {
  constructor(state) {
    if (state) {
      this.state = state;
    } else {
      this.state = new Uint8Array(STATE_SIZE);
      this.state[NEXT_SYMBOL_STATE_INDEX] = X_STATE_VALUE;
      this.state[CONSTRAINT_STATE_INDEX] = UNCONSTRAINED_STATE_VALUE;
    }
    this._legalIndexesSet = null;
  }

  clone() {
    return new UltimateTicTacToe(this.state.slice());
  }

  isEqualTo(uttt) {
    for (let i = 0; i < STATE_SIZE; i += 1) {
      if (this.state[i] !== uttt.state[i]) {
        return false;
      }
    }
    return true;
  }

  execute(action, verify = true) {
    if (verify) {
      if (this.isTerminated()) {
        throw new UltimateTicTacToeError('supergame is terminated');
      }
      this.verifyState();
      this.verifyAction(action);
    }
    this.state[action.index] = action.symbol;
    this.updateSupergameResult(action.symbol, action.index);
    this.toggleNextSymbol();
    this.setNextConstraint(action.index);
    if (verify) {
      this.verifyState();
    }
    this._legalIndexesSet = null;
  }

  getLegalActions() {
    return this.getLegalIndexes().map((legalIndex) => new Action(this.nextSymbol, legalIndex));
  }

  getLegalIndexes() {
    if (this.isTerminated()) {
      return [];
    }
    let indexes;
    if (this.isUnconstrained()) {
      indexes = [];
      for (let subgame = 0; subgame < 9; subgame += 1) {
        if (!this.state[81 + subgame]) {
          indexes = indexes.concat(this.getEmptyIndexes(subgame));
        }
      }
    } else {
      indexes = this.getEmptyIndexes(this.constraint);
    }
    return indexes;
  }

  legalIndexesHas(index) {
    if (!this._legalIndexesSet) {
      this._legalIndexesSet = new Set(this.getLegalIndexes());
    }
    return this._legalIndexesSet.has(index);
  }

  get nextSymbol() {
    return this.state[NEXT_SYMBOL_STATE_INDEX];
  }

  get constraint() {
    return this.state[CONSTRAINT_STATE_INDEX];
  }

  get result() {
    return this.state[UTTT_RESULT_STATE_INDEX];
  }

  get depthLevel() {
    let depthLevel = 0;
    for (let index = 0; index < 81; index++) {
      depthLevel += this.state[index] !== 0;
    }
    return depthLevel;
  }

  isNextSymbolX() {
    return this.nextSymbol === X_STATE_VALUE;
  }

  isNextSymbolO() {
    return this.nextSymbol === O_STATE_VALUE;
  }

  get nextSymbolString() {
    if (this.isNextSymbolX()) {
      return 'x';
    } else if (this.isNextSymbolO()) {
      return 'o';
    }
  }

  isConstrained() {
    return this.constraint >= 0 && this.constraint < 9;
  }

  isUnconstrained() {
    return this.constraint === UNCONSTRAINED_STATE_VALUE;
  }

  isTerminated() {
    return Boolean(this.result);
  }

  isResultX() {
    return this.result === X_STATE_VALUE;
  }

  isResultO() {
    return this.result === O_STATE_VALUE;
  }

  isResultDraw() {
    return this.result === DRAW_STATE_VALUE;
  }

  isSubgameTerminated(subgame) {
    return Boolean(this.state[81 + subgame]);
  }

  isSubgameResultX(subgame) {
    return this.state[81 + subgame] === X_STATE_VALUE;
  }

  isSubgameResultO(subgame) {
    return this.state[81 + subgame] === O_STATE_VALUE;
  }

  isSubgameResultDraw(subgame) {
    return this.state[81 + subgame] === DRAW_STATE_VALUE;
  }

  getEmptyIndexes(subgame) {
    const indexes = [];
    const offset = subgame * 9;
    for (let index = offset; index < offset + 9; index += 1) {
      if (!this.state[index]) {
        indexes.push(index);
      }
    }
    return indexes;
  }

  isWinningPosition(symbol, subgame) {
    const state = this.state;
    const offset = subgame * 9;
    return (
      (symbol === state[offset + 4] &&
        ((symbol === state[offset + 0] && symbol === state[offset + 8]) ||
          (symbol === state[offset + 2] && symbol === state[offset + 6]) ||
          (symbol === state[offset + 1] && symbol === state[offset + 7]) ||
          (symbol === state[offset + 3] && symbol === state[offset + 5]))) ||
      (symbol === state[offset + 0] &&
        ((symbol === state[offset + 1] && symbol === state[offset + 2]) ||
          (symbol === state[offset + 3] && symbol === state[offset + 6]))) ||
      (symbol === state[offset + 8] &&
        ((symbol === state[offset + 2] && symbol === state[offset + 5]) ||
          (symbol === state[offset + 6] && symbol === state[offset + 7])))
    );
  }

  isFull(subgame) {
    const offset = subgame * 9;
    for (let index = offset; index < offset + 9; index += 1) {
      if (!this.state[index]) {
        return false;
      }
    }
    return true;
  }

  updateSupergameResult(symbol, index) {
    let subgameUpdated = false;
    const subgame = Math.trunc(index / 9);
    if (this.isWinningPosition(symbol, subgame)) {
      this.state[81 + subgame] = symbol;
      subgameUpdated = true;
    } else if (this.isFull(subgame)) {
      this.state[81 + subgame] = DRAW_STATE_VALUE;
      subgameUpdated = true;
    }
    if (subgameUpdated) {
      if (this.isWinningPosition(symbol, 9)) {
        this.state[UTTT_RESULT_STATE_INDEX] = symbol;
      } else if (this.isFull(9)) {
        this.state[UTTT_RESULT_STATE_INDEX] = DRAW_STATE_VALUE;
      }
    }
  }

  toggleNextSymbol() {
    if (this.isNextSymbolX()) {
      this.state[NEXT_SYMBOL_STATE_INDEX] = O_STATE_VALUE;
    } else if (this.isNextSymbolO()) {
      this.state[NEXT_SYMBOL_STATE_INDEX] = X_STATE_VALUE;
    }
  }

  setNextConstraint(index) {
    const nextSubgame = index % 9;
    if (this.state[81 + nextSubgame]) {
      this.state[CONSTRAINT_STATE_INDEX] = UNCONSTRAINED_STATE_VALUE;
    } else {
      this.state[CONSTRAINT_STATE_INDEX] = nextSubgame;
    }
  }

  verifyState() {
    this.verifySupergame();
    this.verifySubgames();
    this.verifyConstraint();
  }

  verifySupergame() {
    const x_w = this.isWinningPosition(X_STATE_VALUE, 9);
    const o_w = this.isWinningPosition(O_STATE_VALUE, 9);
    const full = this.isFull(9);
    if (x_w && o_w) {
      throw new UltimateTicTacToeError('X and O have winning positions on supergame');
    }
    if (x_w && !this.isResultX()) {
      throw new UltimateTicTacToeError('X won supergame, but result is not updated');
    }
    if (o_w && !this.isResultO()) {
      throw new UltimateTicTacToeError('O won supergame, but result is not updated');
    }
    if (full && !this.isResultDraw() && !(x_w || o_w)) {
      throw new UltimateTicTacToeError('DRAW on supergame, but result is not updated');
    }
  }

  verifySubgames() {
    for (let subgame = 0; subgame < 9; subgame += 1) {
      const x_w = this.isWinningPosition(X_STATE_VALUE, subgame);
      const o_w = this.isWinningPosition(O_STATE_VALUE, subgame);
      const full = this.isFull(subgame);
      if (x_w && o_w) {
        throw new UltimateTicTacToeError(`X and O have winning positions on subgame=${subgame}`);
      }
      if (x_w && this.state[81 + subgame] !== X_STATE_VALUE) {
        throw new UltimateTicTacToeError(`X won subgame=${subgame}, but supergame is not updated`);
      }
      if (o_w && this.state[81 + subgame] !== O_STATE_VALUE) {
        throw new UltimateTicTacToeError(`O won subgame=${subgame}, but supergame is not updated`);
      }
      if (full && this.state[81 + subgame] !== DRAW_STATE_VALUE && !(x_w || o_w)) {
        throw new UltimateTicTacToeError(`DRAW on subgame=${subgame}, but supergame is not updated`);
      }
    }
  }

  verifyConstraint() {
    if (!(this.isConstrained() || this.isUnconstrained())) {
      throw new UltimateTicTacToeError(`invalid constraint=${this.constraint}`);
    }
    if (this.isConstrained() && this.state[81 + this.constraint]) {
      throw new UltimateTicTacToeError(`constraint=${this.constraint} points to terminated subgame`);
    }
  }

  verifyAction(action) {
    const illegalAction = `Illegal ${action} - `;
    if (this.isNextSymbolX() && !action.isSymbolX()) {
      throw new UltimateTicTacToeError(illegalAction + 'next move belongs to X');
    }
    if (this.isNextSymbolO() && !action.isSymbolO()) {
      throw new UltimateTicTacToeError(illegalAction + 'next move belongs to O');
    }
    if (!(action.index >= 0 && action.index < 81)) {
      throw new UltimateTicTacToeError(illegalAction + 'index outside the valid range');
    }
    if (this.isConstrained() && this.constraint !== Math.trunc(action.index / 9)) {
      throw new UltimateTicTacToeError(illegalAction + `violated constraint=${this.constraint}`);
    }
    if (this.state[81 + Math.trunc(action.index / 9)]) {
      throw new UltimateTicTacToeError(illegalAction + 'index from terminated subgame');
    }
    if (this.state[action.index]) {
      throw new UltimateTicTacToeError(illegalAction + 'index is already taken');
    }
  }

  toString() {
    function stateValueToString(s) {
      if (s === X_STATE_VALUE) {
        return 'X';
      }
      if (s === O_STATE_VALUE) {
        return 'O';
      }
      if (s === DRAW_STATE_VALUE) {
        return '=';
      }
      if (s === 0) {
        return '-';
      }
      return '?';
    }
    function mapStateValuesToStringValues(state) {
      const values = [];
      for (let i = 0; i < state.length; i += 1) {
        values.push(stateValueToString(state[i]));
      }
      return values;
    }
    let subgames = mapStateValuesToStringValues(this.state.slice(0, 81));
    let supergame = mapStateValuesToStringValues(this.state.slice(81, 90));
    if (!this.isTerminated()) {
      const legalIndexes = this.getLegalIndexes();
      for (const legalIndex of legalIndexes) {
        subgames[legalIndex] = '•';
      }
      if (this.isConstrained()) {
        supergame[this.constraint] = '•';
      } else if (this.isUnconstrained()) {
        supergame = supergame.map((s) => (s === '-' ? '•' : s));
      }
    }
    const sb = (l, r) => subgames.slice(l, r + 1).join(' ');
    const sp = (l, r) => supergame.slice(l, r + 1).join(' ');
    subgames = [
      '    0 1 2   3 4 5   6 7 8',
      `  0 ${sb(0, 2)} │ ${sb(9, 11)} │ ${sb(18, 20)}`,
      `  1 ${sb(3, 5)} │ ${sb(12, 14)} │ ${sb(21, 23)}`,
      `  2 ${sb(6, 8)} │ ${sb(15, 17)} │ ${sb(24, 26)}`,
      `    ${'—'.repeat(21)}`,
      `  3 ${sb(27, 29)} │ ${sb(36, 38)} │ ${sb(45, 47)}`,
      `  4 ${sb(30, 32)} │ ${sb(39, 41)} │ ${sb(48, 50)}`,
      `  5 ${sb(33, 35)} │ ${sb(42, 44)} │ ${sb(51, 53)}`,
      `    ${'—'.repeat(21)}`,
      `  6 ${sb(54, 56)} │ ${sb(63, 65)} │ ${sb(72, 74)}`,
      `  7 ${sb(57, 59)} │ ${sb(66, 68)} │ ${sb(75, 77)}`,
      `  8 ${sb(60, 62)} │ ${sb(69, 71)} │ ${sb(78, 80)}`,
    ];
    supergame = ['  ' + sp(0, 2), '  ' + sp(3, 5), '  ' + sp(6, 8)];
    subgames = subgames.join('\n');
    supergame = supergame.join('\n');
    const nextSymbol = stateValueToString(this.nextSymbol);
    const constraint = this.isUnconstrained() ? 'None' : String(this.constraint);
    let result = 'None';
    if (this.isResultX()) {
      result = 'X_WON';
    } else if (this.isResultO()) {
      result = 'O_WON';
    } else if (this.isResultDraw()) {
      result = 'DRAW';
    }
    let output = 'UltimateTicTacToe(\n';
    output += `  subgames:\n${subgames}\n`;
    if (!this.isTerminated()) {
      output += `  next_symbol: ${nextSymbol}\n`;
      output += `  constraint: ${constraint}\n`;
    }
    output += `  supergame:\n${supergame}\n`;
    output += `  result: ${result}\n)`;
    return output;
  }
}
